pkg/controllers/networkpolicy/policy_controller.go

```diff
@@ -0,0 +1,241 @@
+package networkpolicy
+
+import (
+	calicocache "github.com/projectcalico/k8s-policy/pkg/cache"
+	"github.com/projectcalico/k8s-policy/pkg/controllers/controller"
+	"github.com/projectcalico/k8s-policy/pkg/converter"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/fields"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"reflect"
+	"strings"
+	"time"
+)
+
+// PolicyController Implements Controller interface
+// Responsible for monitoring kubernetes network policies and
+// syncing them to Calico datastore.
+type PolicyController struct {
+	indexer        cache.Indexer
+	informer       cache.Controller
+	calicoObjCache calicocache.ResourceCache
+	calicoClient   *client.Client
+	k8sClientset   *kubernetes.Clientset
+}
+
+// NewPolicyController Constructor for PolicyController
+func NewPolicyController(k8sClientset *kubernetes.Clientset, calicoClient *client.Client) controller.Controller {
+	policyConverter := converter.NewPolicyConverter()
+
+	// Function returns map of policyName:policy stored by policy controller
+	// in datastore.
+	listFunc := func() (map[string]interface{}, error) {
+
+		npMap := make(map[string]interface{})
+
+		// Get all policies from datastore
+		calicoPolicies, err := calicoClient.Policies().List(api.PolicyMetadata{})
+		if err != nil {
+			return npMap, err
+		}
+
+		// Filter out only objects that are written by policy controller
+		for _, policy := range calicoPolicies.Items {
+			policyName := policyConverter.GetKey(policy)
+			if strings.Contains(policyName, "knp.default.") {
```

Comment:
I think this needs to be if the string starts with this value, rather than contains this value.

Reply:
For back compatibility, I think we also need to filter in any objects with name of the form `namespace.npName`.


e.g.

```go
if len(strings.Split(policyName, ".")) == 2 {
    // Check k8s cache to see if such a policy exists in the namespace.
    . . .

    // If it does, add it to the map.
    if policyExists {
        npMap[policyName] = policy
    }
}
```

Does that make sense?

Reply:
Grr, scratch that suggestion for now.

I don't think there's a good way to guarantee we'll be backwards compatible, so I'll think about this some more.  LMK if you see a good way to do this...